packages/wallet/src/TxBuilder/buildTx.ts

```diff
@@ -0,0 +1,171 @@
+import { Cardano, NotImplementedError } from '@cardano-sdk/core';
+import { firstValueFrom } from 'rxjs';
+
+import {
+  DeepPartial,
+  MaybeValidTx,
+  OutputBuilder,
+  OutputValidationMinimumCoin,
+  OutputValidationTokenBundleSize,
+  SignedTx,
+  TxBodyValidationError,
+  TxBuilder,
+  TxOutValidationError
+} from './types';
+import { InitializeTxResult, ObservableWallet } from '../types';
+import { OutputBuilderSaw, toOutputValidationError } from './OutputBuilder';
+import { OutputValidator, StakeKeyStatus, createWalletUtil, deepEquals } from '../services';
+import { SignTransactionOptions } from '../KeyManagement';
+
+/**
+ * Minimal sub-type of ObservableWallet that is used by TxBuilder
+ */
+export type TxBuilderObservableWallet = Pick<
+  ObservableWallet,
+  'utxo' | 'protocolParameters$' | 'finalizeTx' | 'initializeTx' | 'submitTx' | 'delegation'
+>;
+
+interface CreateSignedTxProps {
+  wallet: TxBuilderObservableWallet;
+  initializeTxResult: InitializeTxResult;
+  signOptions?: SignTransactionOptions;
+  auxiliaryData?: Cardano.AuxiliaryData;
+}
+
+const createSignedTx = async ({
+  wallet,
+  initializeTxResult,
+  signOptions,
+  auxiliaryData
+}: CreateSignedTxProps): Promise<SignedTx> => {
+  const tx = await wallet.finalizeTx({ auxiliaryData, props: initializeTxResult, signOptions });
+  return {
+    submit: () => wallet.submitTx(tx),
+    tx
+  };
+};
+
+class TxBuilderSaw implements TxBuilder {
+  partialTxBody: Partial<Cardano.NewTxBodyAlonzo> = {};
+  auxiliaryData?: Cardano.AuxiliaryData;
+  extraWitness?: Partial<Cardano.Witness>;
+
+  #observableWallet: TxBuilderObservableWallet;
+  #util: OutputValidator;
+
+  constructor(observableWallet: TxBuilderObservableWallet, util: OutputValidator) {
+    this.#observableWallet = observableWallet;
+    this.#util = util;
+  }
+
+  addOutput(txOut: Cardano.TxOut): TxBuilder {
+    this.partialTxBody.outputs = [...(this.partialTxBody.outputs || []), txOut];
```

Comment:
Maybe let's make it never mutate the `partialTxBody` object too but always recreate as in OutputBuilder?